api/src/lib/ai/ollama.ts

```diff
@@ -0,0 +1,173 @@
+}) {
+  logger.debug(
+    "Generating request data with Ollama",
+    `model: ${model}`,
+    `baseUrl: ${baseUrl}`,
+    `persona: ${persona}`,
+    `method: ${method}`,
+    `path: ${path}`,
+    `handler: ${handler}`,
+    // `handlerContext: ${handlerContext}`,
+    // `openApiSpec: ${openApiSpec}`,
+    // `middleware: ${middleware}`,
+    // `middlewareContext: ${middlewareContext}`,
+  );
+  // Remove trailing slash from baseUrl
+  let openaiCompatibleBaseUrl = baseUrl.replace(/\/$/, "");
+  openaiCompatibleBaseUrl = baseUrl.endsWith("/v1") ? baseUrl : `${baseUrl}/v1`;
+
+  const openaiClient = new OpenAI({ apiKey, baseURL: openaiCompatibleBaseUrl });
+  const userPrompt = await invokeRequestGenerationPrompt({
+    persona,
+    method,
+    path,
+    handler,
+    handlerContext,
+    history,
+    openApiSpec,
+    middleware,
+    middlewareContext,
+  });
+
+  const response = await openaiClient.chat.completions.create({
+    // NOTE - llama-3.1-8b is not very good at function calling, but the 70b and 405b variants were pre-trained to be good at it.
+    //        Don't expect a lot from this :grimace:
+    model,
+    // NOTE - We can try to restrict the response to be from this single tool call
+    tool_choice: {
+      type: "function",
+      function: { name: makeRequestTool.function.name },
+    },
+    tools: [makeRequestTool],
+    messages: [
+      {
+        role: "system",
+        content: getSystemPrompt(persona, "ollama"),
+      },
+      {
+        role: "user",
+        content: userPrompt,
+      },
+    ],
+    temperature: 0.04,
+    max_tokens: 2048,
+  });
+
+  const {
+    choices: [{ message }],
+  } = response;
+
+  logger.debug(
+    "ollama message from response:",
+    JSON.stringify(message, null, 2),
+  );
+
+  const makeRequestCall = message.tool_calls?.[0];
+  const toolArgs = makeRequestCall?.function?.arguments;
+
+  logger.debug("ollama toolArgs", JSON.stringify(toolArgs, null, 2));
+
+  // HACK - llama 3.1 (8B) might get confused and just respond with JSON describing the tool call.
+  //        So we should try to parse the content as a tool call
+  //        ((this does not really work as a fallback yet, but usually the json is malformed anyhow))
+  if (!toolArgs) {
+    logger.warn(
+      chalk.red(
+        "ollama tool call arguments not found, trying to parse message content as tool call",
+      ),
+    );
+    logger.debug("ollama message content", message?.content);
+    const parsedContent = isJson(message?.content)
+      ? JSON.parse(message?.content ?? "")
+      : null;
+    logger.debug("ollama parsedContent", parsedContent);
+    // TODO - Validate the parsed content as a tool call, this object could take a couple of different shapes
+    return parsedContent;
+  }
+
+  try {
+    const parsedArgs = toolArgs ? JSON.parse(toolArgs) : null;
+    // HACK - llama 3.1 does not deal well with nested json objects, and will return each property as its stringified value.
```

Comment:
lots of hacks here to get the function call parameters into the right shape. 

the crux of it is: llama 3.1 8b is not good at respecting the json schema, so a lot of the times it will return nested json stringified params

we do a little hacky magic here to detect if a certain function parameter needs to have its value json parsed